optuna/integration/lightgbm.py

```diff
@@ -62,60 +63,85 @@ class LightGBMPruningCallback(object):
+            # Note that `5` is not the number of folds but the length of sequence.
+            is_cv = len(env.evaluation_result_list) > 0 and len(env.evaluation_result_list[0]) == 5
+            if is_cv:
+                target_valid_name = "cv_agg"
+            else:
+                target_valid_name = self._valid_name
+
+            evaluation_result = self._find_evaluation_result(target_valid_name, env)
+
+            if evaluation_result is None:
+                raise ValueError(
+                    'The entry associated with the validation name "{}" and the metric name "{}" '
+                    "is not found in the evaluation result list {}.".format(
+                        target_valid_name, self._metric, str(env.evaluation_result_list)
+                    )
+                )
+
+            valid_name, metric, current_score, is_higher_better = evaluation_result[:4]
+
             if is_higher_better:
                 if self._trial.study.direction != optuna.study.StudyDirection.MAXIMIZE:
                     raise ValueError(
-                        "The intermediate values are inconsistent with the objective values in "
-                        "terms of study directions. Please specify a metric to be minimized for "
-                        "LightGBMPruningCallback."
+                        "The intermediate values are inconsistent with the objective values"
+                        "in terms of study directions. Please specify a metric to be minimized"
+                        "for LightGBMPruningCallback."
                     )
             else:
                 if self._trial.study.direction != optuna.study.StudyDirection.MINIMIZE:
                     raise ValueError(
-                        "The intermediate values are inconsistent with the objective values in "
-                        "terms of study directions. Please specify a metric to be maximized for "
-                        "LightGBMPruningCallback."
+                        "The intermediate values are inconsistent with the objective values"
+                        "in terms of study directions. Please specify a metric to be"
+                        "maximized for LightGBMPruningCallback."
                     )
 
             self._trial.report(current_score, step=env.iteration)
-            if self._trial.should_prune():
-                message = "Trial was pruned at iteration {}.".format(env.iteration)
-                raise optuna.TrialPruned(message)
-
-            return None
-
-        raise ValueError(
-            'The entry associated with the validation name "{}" and the metric name "{}" '
-            "is not found in the evaluation result list {}.".format(
-                target_valid_name, self._metric, str(env.evaluation_result_list)
-            )
-        )
+
+        if self._trial.should_prune():
+            message = "Trial was pruned at iteration {}.".format(env.iteration)
+            raise optuna.TrialPruned(message)
```

Comment:
Calling `should_prune` seems like it might also slow down performance, and without a `report`, might not have enough data to prune anyway. Did you consider putting this in scope of the if statement above?

Reply:
I considered it, but there are pros/cons.

pros
- we can easily reduce the cost of `should_prune`

cons
- If I configured as `pruning_interval=7` and `reporting_interval=5`, the first time that `should_prune` will be called at 7 x 5 = 91 loops.

Some practical solutions will be:

1. Ask users to use the reasonable intervals for pruning and reporting e.g. `pruning_interval=8` and `reporting_interval=4`.
2. Refactor pruners to minimize the cost of `should_prune`
    - Move [these lines](https://github.com/optuna/optuna/blob/master/optuna/pruners/_percentile.py#L183-L186) to the beginning of `prune`.

Reply:
That feels weird having different intervals for pruning and reporting. Isn't the sole purpose of reporting to prune? What is the use of reporting if not to prune?

Reply:
I heard that some users use report functionality to check intermediate results later.

Reply:
> cons
> If I configured as pruning_interval=7 and reporting_interval=5, the first time that should_prune will be called at 7 x 5 = 35 loops.
> Some practical solutions will be:

Sorry, my previous explanation was wrong. For example, if pruning_interval=10 and reporting_interval=3, pruners will potentially prune the trial at 0, 12, not at the 30 loops.

![image](https://user-images.githubusercontent.com/1277089/116806313-0e9bba00-ab67-11eb-8e66-92fdd44bd278.png)

So, in that sense, your suggestion will have few cons. I will choose your suggestion.

Reply:
> Did you consider putting this in scope of the if statement above?

At first, I was a bit suspicious of the approach, but later I found that the change didn't affect the behavior of the pruners. This is mainly because the pruners check `trial.last_step` to fetch the latest intermediate values, but they do not change if we call `should_prune` without `report`. 

PercentilePruner (MedianPruner): https://github.com/optuna/optuna/blob/7c889c4a8c46cf8c8adc5a967472c0b4900bb428/optuna/pruners/_percentile.py#L188

SuccesiveHalvingPruner (HyperbandPruner):
https://github.com/optuna/optuna/blob/7c889c4a8c46cf8c8adc5a967472c0b4900bb428/optuna/pruners/_successive_halving.py#L165

ThresholdPruner
https://github.com/optuna/optuna/blob/7c889c4a8c46cf8c8adc5a967472c0b4900bb428/optuna/pruners/_threshold.py#L116

In case of `PatientPruner`, the list of `steps` is not changed if we do not call `report`, and `should_prune` keeps the same return values as the last call.
https://github.com/optuna/optuna/blob/7c889c4a8c46cf8c8adc5a967472c0b4900bb428/optuna/pruners/_patient.py#L88




I checked the actual pruning timing of `HyperbandPruner` and `MedianPruner` in the following notebook:
https://colab.research.google.com/drive/14e4H2aWkPGFBQgnOHDq9xmjMjVhnZ-LS?usp=sharing

> Is an interval for pruning even implemented in lightgbm?

I don't think so. We use the callback feature of lightgbm, and it calls the callbacks every iteration.